src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,234 @@
+  extends chisel3.core.ChiselAnnotation
+    with RunFirrtlTransform {
+
+  if(! Seq("h","b").contains(hexOrBinary)) {
+    Builder.error(
+      s"""LoadMemory from file $fileName format must be "h" or "b" not "$hexOrBinary" """
+    )
+  }
+
+  if(fileName.isEmpty) {
+    Builder.warning(
+      s"""LoadMemory from file annotations file empty file name"""
+    )
+  }
+  val SuffixRegex: Regex = """.+\.[^\.]*""".r
+  fileName match {
+    case SuffixRegex(_*) =>
+      Builder.warning(
+        s"""LoadMemoryAnnotation fileName "$fileName" has extension ".txt" will still be appended"""
+      )
+    case _ =>
+  }
+
+  def transformClass : Class[LoadMemoryTransform] = classOf[LoadMemoryTransform]
+
+  def toFirrtl: LoadMemoryAnnotation = LoadMemoryAnnotation(target.toNamed, fileName, hexOrBinary)
+}
+
+/**
+  * Firrtl implementation for load memory
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb
+  */
+case class LoadMemoryAnnotation(
+  target: Named,
+  fileName: String,
+  hexOrBinary: String = "h",
+  originalMemoryNameOpt: Option[String] = None
+) extends SingleTargetAnnotation[Named] {
+
+  def getFileName: String = {
+    (target, originalMemoryNameOpt) match {
+      case (ComponentName(componentName, _), Some(originalMemoryName)) =>
+        if(componentName == originalMemoryName) {
+          fileName
+        }
+        else {
+          fileName + componentName.drop(originalMemoryName.length)
+        }
+      case _ =>
+        fileName
+    }
+  }
+
+  def duplicate(n: Named): LoadMemoryAnnotation = {
+//    if(n.serialize != target.serialize) {
+////      Builder.errors.error(s"LoadMemory[${target.serialize}] must be a simple memory, no bundles")
+//      throw new Exception(s"LoadMemory[${target.serialize}] must be a simple memory, no bundles")
+//    }
+    target match {
+      case ComponentName(componentName, _) =>
+        this.copy(target = n, originalMemoryNameOpt = Some(componentName))
+      case _ =>
+        this.copy(target = n)
+    }
+  }
+}
+
+/**
+  * This pass creates BlackBoxInlineAnno from the LoadMemoryAnnotations
+  * it does this even if the backend is not verilog.
+  *
+  * @param circuitState the target circuit state
+  */
+//TODO: (chick) support a treadle or interpreter means of memory loading
+//TODO: (chick) can this only be done when backend is known to support this.
+//scalastyle:off method.length cyclomatic.complexity regex
+class CreateBindableMemoryLoaders(circuitState: CircuitState) extends Pass {
+
+  val annotations      : Seq[Annotation] = circuitState.annotations
+  val memoryAnnotations: Seq[LoadMemoryAnnotation] = annotations.collect{ case m: LoadMemoryAnnotation => m }
+
+  val bindModules      : mutable.ArrayBuffer[BlackBoxInlineAnno] = new mutable.ArrayBuffer()
+
+  val verilogEmitter   : VerilogEmitter = new VerilogEmitter
```

Comment:
I got this from intellij, it's how their auto formatter does it.
I don't have strong opinions on this, but I tend to defer to them because
* I think there scala offering is strong.
* automatic tooling for this kind of stuff means less typing for me.
This is probably worth a mention in a chisel meeting. I tend to pick long names which exacerbates class and method signatures moving to multiple lines

Reply:
IntelliJ's offering is strong but frequently their auto-formatting runs contrary to published Scala style. For example: https://stackoverflow.com/questions/29405887/how-to-configure-code-style-for-scala-in-intellij-idea

Most of these things are configurable, I don't feel particularly strongly about this particular one although I'm more inclined toward what @seldridge suggested. I suspect we should fix some of the defaults (like the one in the linked StackOverflow post) and provide a template.

Reply:
Yeah, I'm not advocating to have this hold it up. Only an alternative formatting to consider. I'm usually defaulting to whatever Emacs' `scala-mode` does for me and then using some quick regex alignment/macros to clean up stuff like this if I have trouble reading it. Sometimes `scala-mode` will do some dumb formatting, too, and I have to coax it back to normalcy. However, it's likely easier for me to deal with these types of alignments than others. Not expecting that. Fundamentally, if the style checks sign off, I'm fine. Long names are totally fine, too. I like descriptive names unless they're one-time-use and/or very brief scope/utility, e.g., [extractor temporaries](https://github.com/freechipsproject/firrtl/pull/765/files#diff-57408ffab8e042ff36bc22f6cad28c27R56).

Reply:
An alternative is that we could try is [scalafmt](https://scalameta.org/scalafmt/) such that no editors defaults matter and everything is consistent. Of course, I worry that such an approach would not mesh well with some non-standard things where we want different formatting ([eg. rocket-chip's decoder](https://github.com/freechipsproject/rocket-chip/blob/53a162c826fd4298766ba2b67154937e27231e3e/src/main/scala/rocket/IDecode.scala#L49))